src/kbmod/search/raw_image_eigen.cpp

```diff
@@ -0,0 +1,767 @@
+    if  (image(itl.i, itl.j) != NO_DATA)
+      interpolated_value += w_tl * image(itl.i, itl.j);
+
+    if (image(itr.i, itr.j) != NO_DATA)
+      interpolated_value += w_tr * image(itr.i, itr.j);
+
+    if (image(ibl.i, ibl.j) != NO_DATA)
+      interpolated_value += w_bl * image(ibl.i, ibl.j);
+
+    if (image(ibr.i, ibr.j) != NO_DATA)
+      interpolated_value += w_br * image(ibr.i, ibr.j);
+
+    if (interpolated_value == 0.0f)
+      return NO_DATA;
+    return interpolated_value;
+  }
+
+
+  RawImageEigen RawImageEigen::create_stamp(const float x, const float y, const int radius,
+                                            const bool interpolate, const bool keep_no_data) const {
+    if (radius < 0)
+      throw std::runtime_error("stamp radius must be at least 0");
+
+    const int dim = radius * 2 + 1;
+    Index idx(x, y, false);
+    // evaluates as rvalue. The constructor assigns the block to an array, type
+    // Image, prompting the evaluation of the expression
+    // to fix the constness issue we will materialize a copy of the array
+    Image block = image.block(idx.i, idx.j, dim, dim);
+    RawImageEigen stamp(block, obstime);
+
+    if (interpolate) {
+      for (int yoff = 0; yoff < dim; ++yoff) {
+        for (int xoff = 0; xoff < dim; ++xoff) {
+          // double check if it was this->interpolate or stamp.interpolate
+          stamp.image(yoff, xoff) = this->interpolate(x + static_cast<float>(xoff - radius),
+                                                      y + static_cast<float>(yoff - radius));
+        }
+      }
+    }
+
+    return stamp;
+  }
+
+
+  void RawImageEigen::add(const float x, const float y, const float value) {
+    // In the original implementation this just does array[x, y] += value
+    // even though that doesn't really make any sense? The cast is probably
+    // implicit. This is likely slower than before, because the floor is used
+    // not a cast. Floor is more correct, but I added the constructor nonetheless
+    Index p(x, y, false);
+    if (p.is_within(width, height))
+      image(p.i, p.j) += value;
+  }
+
+
+  void RawImageEigen::add(const unsigned x, const unsigned y, const float value) {
+    Index p(x, y);
+    if (p.is_within(width, height))
+      image(p.i, p.j) += value;
+  }
+
+
+  void RawImageEigen::add(const Index p, const float value) {
+    if (p.is_within(width, height))
+      image(p.i, p.j) += value;
+  }
+
+
+  void RawImageEigen::interpolated_add(const float x, const float y, const float value) {
+    // interpolation weights and neighbors
+    auto iwn = get_interp_neighbors_and_weights(x, y);
+    add(iwn.top_left,  value * iwn.w_tl);
+    add(iwn.top_right, value * iwn.w_tr);
+    add(iwn.bot_left,  value * iwn.w_bl);
+    add(iwn.bot_right, value * iwn.w_br);
+  }
+
+
+  std::array<float, 2> RawImageEigen::compute_bounds() const {
+    float min_val = FLT_MAX;
+    float max_val = -FLT_MAX;
+
+    for (auto elem : image.reshaped())
```

Comment:
I'm hoping there is a more vectorized approach that we can use here in the future. No need to change it now though.

Reply:
I worked out the solution but it requires implementing the `minCoeff` and `maxCoeff` from scratch by invoking the vectorized comparison operators in the `Eigen::internal`. Very messy. This would be a lot easier to hand over to the Eigen vectorized internals if we kept a separate bitmask, set the values to `FLT_MAX`, `FLT_MIN` and call regular `minCoeff` and `maxCoeff` implementations respectively.

A problem with masked min/max methods is that it's it's not a well defined operations, what happens when, f.e., all array values are masked. 

Here is the sample for just one of the two functions, tested to work and vectorize with SSE. This is well into the internals of Eigen, isn't well documented, and quite likely isn't as good as it should be. Whether or not the vectorized impl is called is very circumstantial, most of the time it isn't. 

```
// this is the internal boolean reduction operation
// https://eigen.tuxfamily.org/dox/group__TutorialReductionsVisitorsBroadcasting.html
// https://gitlab.com/libeigen/eigen/-/blob/master/Eigen/src/Core/functors/BinaryFunctors.h#L136
template<typename LhsScalar,typename RhsScalar, typename MaskScalar, int NaNPropagation>
struct masked_scalar_min_op : Eigen::internal::binary_op_base<LhsScalar,RhsScalar>
{
  masked_scalar_min_op(LhsScalar mask_val) : mask(mask_val) {};

  typedef typename Eigen::ScalarBinaryOpTraits<LhsScalar, RhsScalar, masked_scalar_min_op>::ReturnType result_type;
  MaskScalar mask;

  inline result_type operator() (const LhsScalar& a, const RhsScalar& b) const {
    // I'm afraid that I'm not fully sure what comparison operator to use
    return Eigen::internal::pmin(Eigen::internal::pselect(Eigen::internal::pcmp_eq(a, mask), b, a),
                                 Eigen::internal::pselect(Eigen::internal::pcmp_eq(b, mask), a, b));
  }

  //It doesn't seem like the comparison operators are vectorized by default, there must be a
  // flag somewhere. See
  // https://gitlab.com/libeigen/eigen/-/blob/master/Eigen/src/Core/GenericPacketMath.h#L584
  template<typename Packet>
  inline Packet packetOp(const Packet& a, const Packet& b) const
  {
    Packet no_val_a = Eigen::internal::pcmp_eq(a, Eigen::internal::pset1(mask));
    Packet no_val_b = Eigen::internal::pcmp_eq(b, Eigen::internal::pset1(mask));
    Packet not_nan_mask_a = pcmp_eq(a, a);
    Packet not_nan_mask_b = pcmp_eq(b, b);
    Packet not_mask_a = Eigen::internal::por(no_val_a, not_nan_mask_a);
    Packet not_mask_b = Eigen::internal::por(no_val_b, not_nan_mask_b);
    return pselect(not_mask_a,
                           pselect(not_mask_b, op(a, b), b),
                           a);
  }

  // This is mentioned as the slow generic implementation of redux
  // so Imma guess I won't be touching this, and this is for more complicated
  // matrices with like CMPLX numbers in them etc. Technically we should provide
  // an implementation here too.
  template<typename Packet>
  inline result_type predux(const Packet& a) const
  {
    return internal::predux_min<NaNPropagation>(a);
  }

};

// this is the actual function we invoke
template<typename Derived>
inline typename Eigen::internal::traits<Derived>::Scalar maskedMinCoeff(const Eigen::DenseBase<Derived>& d,
                                                                        const double maskval){
  auto res = d.redux(masked_scalar_min_op<
                     typename Eigen::internal::traits<Derived>::Scalar,
                     typename Eigen::internal::traits<Derived>::Scalar,
                     typename Eigen::internal::traits<Derived>::Scalar,
                     Eigen::NaNPropagationOptions::PropagateNumbers
                     > (maskval));

  // this jmp is marked costly
  if (res == maskval) throw std::runtime_error("Min and Max are undefined!");
  return res;
}
```

The example usage would be
```
int main(){
  Eigen::MatrixXd img = Eigen::MatrixXd::Zero(10, 10);
  img(1, 1) = 100;
  img(0, 0) = -100;
  img(1, 2) = -999.0;
  img(1, 3) = 999.0;
  std::cout << maskedMinCoeff(ri.get_image(), -999.0) << " expected -100" << std::endl;
  std::cout << maskedMinCoeff(ri.get_image(), 999.0) << " expected -999" << std::endl;
  std::cout << maskedMinCoeff(ri.get_image(), 0.0) << " expected -999" << std::endl;
}
// -100 expected -100
// -999 expected -999
// -999 expected -999
```

Ergo the preference to move to separate bitmask.